File: GLLara/GLLItem+OBJExport.m
//
//  GLLItem+OBJExport.m
//  GLLara
//
//  Created by Torsten Kammer on 19.09.12.
//  Copyright (c) 2012 Torsten Kammer. All rights reserved.
//

#import "GLLItem+OBJExport.h"

#import "NSArray+Map.h"
#import "GLLItemBone.h"
#import "GLLItemMesh+OBJExport.h"
#import "GLLModelBone.h"
#import "GLLModelMesh.h"

@implementation GLLItem (OBJExport)

- (BOOL)willLoseDataWhenConvertedToOBJ
{
	for (GLLItemMesh *mesh in self.meshes)
		if (mesh.willLoseDataWhenConvertedToOBJ)
			return YES;
	
	return NO;
}

- (NSString *)objStringForFilename:(NSString *)name withTransform:(BOOL)transform withColor:(BOOL)color error:(NSError *__autoreleasing*)error;
{
	NSMutableString *obj = [NSMutableString string];
	
	NSString *materialLibraryName = [[name stringByDeletingPathExtension] stringByAppendingPathExtension:@"mtl"];
	[obj appendFormat:@"mtllib %@\n", materialLibraryName];
	
	mat_float16 *transforms = malloc(sizeof(mat_float16) * self.bones.count);
	NSUInteger boneIndex = 0;
	for (GLLItemBone *bone in self.bones)
	{
		if (transform)
			[bone.globalTransform getValue:&transforms[boneIndex]];
		else
			transforms[boneIndex] = bone.bone.positionMatrix;
		boneIndex += 1;
	}
	
	uint32_t indexOffset = 0;
	for (GLLItemMesh *mesh in self.meshes)
	{
		[obj appendString:[mesh writeOBJWithTransformations:transforms baseIndex:indexOffset includeColors:color]];
		indexOffset += mesh.mesh.countOfVertices;
	}
	
	free(transforms);

	return obj;
}
- (NSString *)mtlStringForLocation:(NSURL *)location error:(NSError *__autoreleasing*)error;
{
	return [[self.meshes map:^(GLLItemMesh *mesh) {
		return [mesh writeMTLWithBaseURL:location];
	}] componentsJoinedByString:@"\n"];
}

- (BOOL)writeOBJToLocation:(NSURL *)location withTransform:(BOOL)transform withColor:(BOOL)color error:(NSError *__autoreleasing*)error;
{
	return [[self objStringForFilename:location.lastPathComponent withTransform:transform withColor:color error:error] writeToURL:location atomically:YES encoding:NSUTF8StringEncoding error:error];
}

- (BOOL)writeMTLToLocation:(NSURL *)location error:(NSError *__autoreleasing*)error;
{
	return [[self mtlStringForLocation:location error:error] writeToURL:location atomically:YES encoding:NSUTF8StringEncoding error:error];
}

@end

File: GLLara/GLLItem+OBJExport.swift
//
//  GLLItem+OBJExport.m
//  GLLara
//
//  Created by Torsten Kammer on 19.09.12.
//  Copyright (c) 2012 Torsten Kammer. All rights reserved.
//

import Foundation

extension GLLItem {
    @objc var willLoseDataWhenConvertedToOBJ: Bool {
        return meshes.first { ($0 as! GLLItemMesh).willLoseDataWhenConvertedToOBJ } != nil
    }
    
    @objc func writeOBJ(location: URL, transform: Bool, color: Bool) throws {
        var string = ""
        
        let materialLibraryName = location.deletingPathExtension().appendingPathExtension(".mtl").lastPathComponent
        string += "usemtl \(materialLibraryName)\r\n"
        
        var transforms = Array<mat_float16>(repeating: mat_float16(), count: bones.count)
        var boneIndex = 0
        for bone in bones {
            if transform {
                transforms[boneIndex] = (bone as! GLLItemBone).globalTransform
            } else {
                transforms[boneIndex] = (bone as! GLLItemBone).bone.positionMatrix
            }
            boneIndex += 1
        }
        
        var indexOffset = 0
        for mesh in meshes {
            string += (mesh as! GLLItemMesh).writeOBJ(transformations: transforms, baseIndex: indexOffset, includeColors: color)
            indexOffset += (mesh as! GLLItemMesh).mesh.countOfVertices
        }
        try string.write(to: location, atomically: true, encoding: .utf8)
    }
    
    @objc func writeMTL(location: URL) throws {
        let content = meshes.map { ($0 as! GLLItemMesh).writeMTL(baseURL: location) }.joined(separator: "\r\n")
        try content.write(to: location, atomically: true, encoding: .utf8)
    }
}
